<a href="https://colab.research.google.com/github/JDevine1981/Prediction-of-Product-Sales/blob/main/Project1Part2Core%2B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## LOADING DATA

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
df = pd.read_csv('/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv')

In [ ]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


# DATA CLEANING

### How many rows/columns?

In [4]:
df.shape

(8523, 12)

### Review Data Types

In [5]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

All columns appear to be assigned the appropriate data type. Further inspection will determine if any columns
will need to be converted.

## Duplicated Data

Checking for Duplicated Data

In [6]:
df.duplicated().sum()

0

There are no duplicated values.

## Unique Values

In [7]:
df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

The column Item_Fat_Content contains five unqiue values, but according to our data dictionary,
it should only contain two: Low-Fat and Regular.

In [8]:
# Reviewing value counts for Item_Fat_Content
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [9]:
# Replacing and Standardizing Inconsistent Values
df["Item_Fat_Content"] = df['Item_Fat_Content'].replace({'low fat':'Low Fat',
'reg':'Regular', 'LF': 'Low Fat'})
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

Checking for missing values

In [10]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Identify % of missing data

In [11]:
df.isna().sum() / len(df) * 100

Item_Identifier               0.000000
Item_Weight                  17.165317
Item_Fat_Content              0.000000
Item_Visibility               0.000000
Item_Type                     0.000000
Item_MRP                      0.000000
Outlet_Identifier             0.000000
Outlet_Establishment_Year     0.000000
Outlet_Size                  28.276428
Outlet_Location_Type          0.000000
Outlet_Type                   0.000000
Item_Outlet_Sales             0.000000
dtype: float64

## Missing Values:

  There are 1,463 missing values for Item_Weight (17%).

  There are 2,410 missing values for Outlet_size (28%).

  This presents an opportunity to review the object columns and decide whether we can fill in the missing data as MISSING, impute with a value, or drop the column altogether.

In [12]:
# Create a filter to use in a for loop to loop through the object type columns
data_types = df.dtypes
str_cols = data_types[data_types=='object'].index
str_cols

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [13]:
# Run the for loop
for col in str_cols:
  print(f'-{col}:')
  print(df[col].value_counts(dropna=False))
  print('\n\n')

-Item_Identifier:
FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64



-Item_Fat_Content:
Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64



-Item_Type:
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64



-Outlet_Identifier:
OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifi

Since we are provided with both the location type and outlet type, and neither column contains missing values, it is safe to assume outlet size is a superfluous column that can be dropped.

In [14]:
# Drop the Outlet Size column from the DataFrame
df = df.drop(columns=['Outlet_Size'])

In [15]:
# Review Changes
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Location_Type        Outlet_Type  \
0                       1999               Tier 1  Supermarket Type1   
1                       2009               Tier 3  Supermarket Type2   
2                       1999               Tier 1  Supermarket Type1   
3                       1998               Tier 3      Grocery Store   
4                       1987               Tier 3  Supermarket Type1   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

We can also rename several columns by removing the terms item and outlet, as it is implied the data refers to the products and locations.

In [16]:
# Create a rename dictionary
rename_dict = {'Item_Weight':'Weight', 'Item_Fat_Content':'Fat_Content', 'Item_Visibility':'Visibility',
'Item_Type':'Type', 'Item_MRP': 'MRP', 'Outlet_Establishment_Year':'Establishment_Year', 'Outlet_Location_Type':
'Location_Type'}
rename_dict

{'Item_Weight': 'Weight',
 'Item_Fat_Content': 'Fat_Content',
 'Item_Visibility': 'Visibility',
 'Item_Type': 'Type',
 'Item_MRP': 'MRP',
 'Outlet_Establishment_Year': 'Establishment_Year',
 'Outlet_Location_Type': 'Location_Type'}

In [17]:
# Dictionary substitution using rename method
df = df.rename(rename_dict, axis=1)
df.head()

Item_Identifier  Weight Fat_Content  Visibility                   Type  \
0           FDA15    9.30     Low Fat    0.016047                  Dairy   
1           DRC01    5.92     Regular    0.019278            Soft Drinks   
2           FDN15   17.50     Low Fat    0.016760                   Meat   
3           FDX07   19.20     Regular    0.000000  Fruits and Vegetables   
4           NCD19    8.93     Low Fat    0.000000              Household   

        MRP Outlet_Identifier  Establishment_Year Location_Type  \
0  249.8092            OUT049                1999        Tier 1   
1   48.2692            OUT018                2009        Tier 3   
2  141.6180            OUT049                1999        Tier 1   
3  182.0950            OUT010                1998        Tier 3   
4   53.8614            OUT013                1987        Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

For the item weight, we will fill in the missing values with the placeholder '-1'. Further inspection of the data may reveal the item weight is already listed in a duplicate row. There are only 1559 unique values for product ID in 8,523 rows, so it is likley the missing item weights are listed in one or more of the duplicate rows.

In [18]:
df['Weight'] = df['Weight'].fillna(-1)
df['Weight'].isna().sum()

0

In [19]:
# Checking describe after filling null values
df['Weight'].describe()

count    8523.000000
mean       10.478936
std         6.720742
min        -1.000000
25%         6.650000
50%        11.000000
75%        16.000000
max        21.350000
Name: Weight, dtype: float64

In [20]:
# Confirm no additional missing values
df.isna().sum()

Item_Identifier       0
Weight                0
Fat_Content           0
Visibility            0
Type                  0
MRP                   0
Outlet_Identifier     0
Establishment_Year    0
Location_Type         0
Outlet_Type           0
Item_Outlet_Sales     0
dtype: int64

Upon further inspection, the establishment year yields little value in determininig items sales based on product and location. We will drop this column as well.

In [21]:
# Drop the establishment_year column from the dataframe
df = df.drop(columns=['Establishment_Year'])

In [22]:
# Review Changes
df.head()

Item_Identifier  Weight Fat_Content  Visibility                   Type  \
0           FDA15    9.30     Low Fat    0.016047                  Dairy   
1           DRC01    5.92     Regular    0.019278            Soft Drinks   
2           FDN15   17.50     Low Fat    0.016760                   Meat   
3           FDX07   19.20     Regular    0.000000  Fruits and Vegetables   
4           NCD19    8.93     Low Fat    0.000000              Household   

        MRP Outlet_Identifier Location_Type        Outlet_Type  \
0  249.8092            OUT049        Tier 1  Supermarket Type1   
1   48.2692            OUT018        Tier 3  Supermarket Type2   
2  141.6180            OUT049        Tier 1  Supermarket Type1   
3  182.0950            OUT010        Tier 3      Grocery Store   
4   53.8614            OUT013        Tier 3  Supermarket Type1   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Item_Identifier    8523 non-null   object 
 1   Weight             8523 non-null   float64
 2   Fat_Content        8523 non-null   object 
 3   Visibility         8523 non-null   float64
 4   Type               8523 non-null   object 
 5   MRP                8523 non-null   float64
 6   Outlet_Identifier  8523 non-null   object 
 7   Location_Type      8523 non-null   object 
 8   Outlet_Type        8523 non-null   object 
 9   Item_Outlet_Sales  8523 non-null   float64
dtypes: float64(4), object(6)
memory usage: 666.0+ KB


SUMMARY STATISTICS FOR NUMERIC VALUES

The minimum value for visibility is 0, which is not possible. As with the other missing values, we will need to either remove the rows with missing data, fill with a placehodler, or impute a value.

In [24]:
df.describe().round(2)

Weight  Visibility      MRP  Item_Outlet_Sales
count  8523.00     8523.00  8523.00            8523.00
mean     10.48        0.07   140.99            2181.29
std       6.72        0.05    62.28            1706.50
min      -1.00        0.00    31.29              33.29
25%       6.65        0.03    93.83             834.25
50%      11.00        0.05   143.01            1794.33
75%      16.00        0.09   185.64            3101.30
max      21.35        0.33   266.89           13086.96

Since 0.00 is not actually a null value, we will first need to convert it to filter all rows with 0.00 visibility and set the value for that filter to null before we can fill in a placeholder.

In [25]:
# Create a filter to identify the rows where Visibility = 0.00
filter_zero_vis = df['Visibility'] == 0.00

In [26]:
# Change to a null value
df.loc[filter_zero_vis, 'Visibility']=np.nan

In [27]:
# Confirm change to null value
df['Visibility'].describe()

count    7997.000000
mean        0.070482
std         0.050308
min         0.003575
25%         0.031403
50%         0.057792
75%         0.098109
max         0.328391
Name: Visibility, dtype: float64

In [28]:
df['Visibility'].isna().sum()

526

In [29]:
# Fill the missing values using -1
df['Visibility'] = df['Visibility'].fillna(-1)

In [30]:
# Confirm changes
df['Visibility'].isna().sum()

0

In [31]:
df.head()

Item_Identifier  Weight Fat_Content  Visibility                   Type  \
0           FDA15    9.30     Low Fat    0.016047                  Dairy   
1           DRC01    5.92     Regular    0.019278            Soft Drinks   
2           FDN15   17.50     Low Fat    0.016760                   Meat   
3           FDX07   19.20     Regular   -1.000000  Fruits and Vegetables   
4           NCD19    8.93     Low Fat   -1.000000              Household   

        MRP Outlet_Identifier Location_Type        Outlet_Type  \
0  249.8092            OUT049        Tier 1  Supermarket Type1   
1   48.2692            OUT018        Tier 3  Supermarket Type2   
2  141.6180            OUT049        Tier 1  Supermarket Type1   
3  182.0950            OUT010        Tier 3      Grocery Store   
4   53.8614            OUT013        Tier 3  Supermarket Type1   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Item_Identifier    8523 non-null   object 
 1   Weight             8523 non-null   float64
 2   Fat_Content        8523 non-null   object 
 3   Visibility         8523 non-null   float64
 4   Type               8523 non-null   object 
 5   MRP                8523 non-null   float64
 6   Outlet_Identifier  8523 non-null   object 
 7   Location_Type      8523 non-null   object 
 8   Outlet_Type        8523 non-null   object 
 9   Item_Outlet_Sales  8523 non-null   float64
dtypes: float64(4), object(6)
memory usage: 666.0+ KB


In [34]:
# Saving the cleaned data to Google Drive
fpath_out = "/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales-predictioncore-eda.csv"
df.to_csv(fpath_out, index=True)

In [37]:
# Loading and inspecting saved file
loaded = pd.read_csv(fpath_out)
loaded.head()
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Item_Identifier    8523 non-null   object 
 1   Weight             8523 non-null   float64
 2   Fat_Content        8523 non-null   object 
 3   Visibility         8523 non-null   float64
 4   Type               8523 non-null   object 
 5   MRP                8523 non-null   float64
 6   Outlet_Identifier  8523 non-null   object 
 7   Location_Type      8523 non-null   object 
 8   Outlet_Type        8523 non-null   object 
 9   Item_Outlet_Sales  8523 non-null   float64
dtypes: float64(4), object(6)
memory usage: 666.0+ KB


Item_Identifier      0
Weight               0
Fat_Content          0
Visibility           0
Type                 0
MRP                  0
Outlet_Identifier    0
Location_Type        0
Outlet_Type          0
Item_Outlet_Sales    0
dtype: int64